hibernate-core/src/main/java/org/hibernate/event/internal/DefaultFlushEntityEventListener.java

```diff
@@ -332,15 +332,8 @@ protected boolean handleInterception(FlushEntityEvent event) {
 		final boolean intercepted = invokeInterceptor( session, entity, entry, values, persister );
 
 		//now we might need to recalculate the dirtyProperties array
-		if ( intercepted && event.isDirtyCheckPossible() && !event.isDirtyCheckHandledByInterceptor() ) {
-			int[] dirtyProperties;
-			if ( event.hasDatabaseSnapshot() ) {
-				dirtyProperties = persister.findModified( event.getDatabaseSnapshot(), values, entity, session );
-			}
-			else {
-				dirtyProperties = persister.findDirty( values, entry.getLoadedState(), entity, session );
-			}
-			event.setDirtyProperties( dirtyProperties );
+		if ( intercepted && event.isDirtyCheckPossible() ) {
```

Comment:
Why not `!event.isDirtyCheckHandledByInterceptor()`? Doesn't that effectively disable the dirty check enhancement?

Reply:
Why would that disable the dirty check enhancement? The logic for which dirty checking mechanism to use is inside `dirtyCheck(final FlushEntityEvent event)` method.

Reply:
I think I agree with @JWGmeligMeyling . In this case, it's the second time you are calling `dirtyCheck()` so you might as well use the flags that are in the `event` to avoid calling `dirtyCheck()` as it has already been decided that dirty checking is using an interceptor.

Or did we miss something and the test fails when you keep ` && !event.isDirtyCheckHandledByInterceptor()`?

Reply:
Exactly. The test fails if we add the check. So, we can either call dirtyChecl or duplicate it's logic. It's not just bytecode enhancement to consider. We need to consider the custom mechanisms and the interceptor as well because the user might use a custom mechanism to determine that the entity was changed by the call back.

Reply:
Thanks for the explanation. Looks good to me, please merge!

I think we should backport it to 5.2.18 too.

Reply:
I'm thinking of adding, in case we don't already have, a test for the custom dirty checking mechanism/Session interceptor to make sure we test that and the logic is covered.

Reply:
I agree this change is correct.